### PR TITLE
0002 datamodel 2

### PR DESCRIPTION
**Pull request by matthewmueller**
Summarized and up-to-date version of DM2 spec in #16. There's still a lot of stuff in #16 that we'll need to bring over but this represents the scope of what we're doing for the initial release.

[Rendered View](https://github.com/prisma/rfcs/blob/5c5da8ab003c690257dc3f72ce267374a64a1d5b/text/0002-datamodel.md)

### CODE REVIEW

**Comment by fleabitten232** [comment]
I'm finding some of the formatting and syntax changes unpleasant enough that it's overcome my desire to not get involved ;)

**Aligning blocks by their = sign**
For what feels like a tiny improvement in readability that mostly applies with you don't have code coloring, you inherit a lot of unpleasantness. 

- Blocks become inconsistent with other blocks and even within themselves if there's a newline. 
- An IDE with autoformat is mandatory if you want to avoid losing hours to carefully aligning everything yet again because you added a field with a slightly longer name. 
- Any if you're assuming there is autoformat available, you can assume there is code coloring available which does *much* more for improving readability without the sacrifices in writeability.

I know the stated reason for enforcing these formatting rules is to avoid bikeshedding but this feels like the *result* of someone spending hours bikeshedding that is now going to be forced upon everyone else.

The "JSON/CSS-ish object with one space between fields/types/directives" was consistent everywhere, always and with proper IDE coloring is more readable and vastly more writable.

**Using an = sign at all**
Is this solving some problem I'm unaware of or is it just trying to make the schema look less like SDL? 
SDL hits a nice balance between JSON/CSS-like without following it strictly enough that it ends up feeling too wordy. 

JSON is about as ubiquitous as you can get for describing a data structure and I can't see any reason to stray away from it without good justification -- *especially* when you're building something that is a part of a web stack.

Maybe there's some justification of all these changes that I haven't found either reading the RFCs or using Prisma 1.x but right now, they're really not sitting well with me.

**Comment by fleabitten232** [comment]
I feel like the formatting has been majorly overthought and that the changes lack any real justification, like one person getting wrapped up in bikeshedding about stopping people from getting wrapped up in bikeshedding. Here are my thoughts and suggestions rolled up together...

### Formatting Rules

**Configuration blocks are align[sic] by their = sign**
- Aligning by = signs is terrible for writeability, 
  - Makes autoformat borderline mandatory for your IDE. What about when you're making a quick change to something over ssh with something like nano? What about before you hit CTRL+S?
 - The spacing is inconsistent between blocks (and even inconsistent *within* blocks with the new line / multiline rules)
- Only slightly increases readability and only in perfect cases -- the multiline objects look significantly worse, a page full of blocks with different spacing looks messy, etc
- Provides worse separation between names/keys/values than code coloring does (a feature that's far more common than auto formatting)

Whats the justification for the format in the first place? JSON for describing configs and objects is as close to ubiqutious as you'll get in the programming world, *especially* when you're working on with web stacks.

So my suggestion: use JSON. It's fine.
```
config: {
  generate: 
    js: {
      target: 'es3',
      snakeCase = true
    }
  },
  connector: {
    pg {
      type = "postgres"
      url  = env.POSTGRES_URL
    }
  }
}
```

**Field definitions are aligned into columns separated by 2 or more spaces.**
 - Suffers from many the problems listed above; tedious mangement of whitespace, only consistent for each block, doesn't add much to readability, the newline reset is messy, etc.
 - Most editors won't properly respect this column-based indentation when adding a new line. For example, typing a directive and then hitting enter to type another will force you to tab your way through two empty columns. Single indentation doesn't have that problem -- hitting enter will move your caret to the correct position and you're only one keypress away from indent/outdenting it.

My rewrite of one of your examples:
```
model User {
  id: String
  name: String
  customer: @embed {
      id: String
      full_name: String
      cards: @embed {
        type: Card
      }[]
  }?
  age: Int
  email: String
  example: String @directive
  another: String
    @withMulti
    @lineDirectives
}
```
One caveat of this formatting is the inconsistency of the directives -- a single directive goes on the same line but multiple directives must have their own line each.

But since part of this is an exercise is breaking away from SDL, this is also quite readable (and would read even better with proper code coloring and indentation hints):
```
model User {
  @@unique([ title, slug ])
  @@attribute1

  id: String
  name: String
  customer: Model 
    @embed {
      id: String
      full_name: String
      cards:  Model
        @embed [{
          type: Card
        }[]
    }?
  age: Int
  email: String
  example: String 
    @directive
  another: String
    @withMulti
    @lineDirectives
}
```

And yet another slightly different example that's my favorite...

```
block name {
  @@unique([title, slug])
  @@attribute1

  keyOne: type
  keyTwo: type @directive
  keyThree: [type]
  keyFour: type
    @thisField
    @hasLots(value: 0)
    @ofDirectives
    @soAllowNewlines
  keyFive: 
    @embed {
      subkeyOne: type @directive
      subkeyTwo: type 
        @directive
        @moreDirective
      subkeyThree: type
    }
  keySix:
    @thankYouForReading 
    @embed {
      subkeyOne: type @directive
      subkeyTwo: type 
        @directive
        @moreDirective
      subkeyThree: type
    }
}
```

I know it's a minor thing to fixate on but one of the things that drew me to Prisma in the first place was its code was clean, pretty and to the point compared to other ORMs and I feel the current suggestions have been changed simply for the sake of change, rather than having any real justification.

**Comment by matthewmueller** [comment]
Hey @fleabitten232, thanks for your thoughtful feedback. I'll try and address each point and then we can discuss further.

> Aligning by = signs is terrible for writeability,

Space alignment is optional, so the syntax still works correctly if you do the following:

```
block fruits {
apple = "apple"
banana = "banana"
pear=5
}

model User {
id Int @id
first_name String @unique
}
```

In fact, the parser doesn't consider spacing at all in decision making.

> The spacing is inconsistent between blocks

We're replacing the 2-space alignment with 1-space alignment to make this more consistent. If your'e seeing other consistencies, please let me know. It's probably a mistake.

> The multiline objects look significantly worse

We'll probably wait on this before adding it to ensure that it's a problem. The main reason it's in the spec is because you don't want really long models shifting everything with a new long column. Go and Terraform's formatters both provide this feature to help with this.

> Provides worse separation between names/keys/values than syntax highlighting

Could you elaborate with an example? I'm not sure I understand :-)

>  use JSON. It's fine.

I'm all for simplicity and not reinventing the wheel. We indeed considered using JSON. The two constraints we're dealing with are that models don't look great this way:

```json
{
  models: {
    user: {
      id: { type: "Int", attributes: ["@id"] },
      first_name: { type: "String", attributes: ["@unique"] }
    }
  }
}
```

And blocks can reference each other:

```
datasource pg {
  connector = "postgres"
  url = "postgres://localhost:5432/db
}

model User {
  id Int @id @pg.uuid()
}
```

We could have took inspiration from JSON schema in how they do `$ref`, but I found that to be a hard-to-read solution.

> And yet another slightly different example that's my favorite...

I quite like that solution, and in fact that's almost valid with our DM2 syntax! From the parser we don't care about whitespace, so if you decide not to use the formatter you could write your @attributes like that and they would apply to the field above.

**Comment by fleabitten232** [comment]
Thanks for the explanation.

> Space alignment is optional

Optional, but if you don't want to use it you'll have to fight the formatter about it. I'm okay with not polluting any formatter with 100 options as long as the base is good ;)

> Could you elaborate with an example? I'm not sure I understand :-)

I find that your examples which have been spaced into columns...
```
model User {
  id        Int        @id
  customer  Customer?  @relation(id)
  name      String
}

model Customer {
  id       Int     @id
  user     User?
  address  String
}
```
Are less readable than just a single space diving them them (as long as you have syntax highlighting that correctly separates name, type and directives with different colors)

formatexample

And the column spacing makes writing everything worse also. We start with this..
```
model User {
  id        Int         @id
  asked     Question[]  @relation(name: "asker")
  answered  Question[]  @relation(name: "answerer")
}

model Question {
  id        Int   @id
  asker     User  @relation(name: "asker")
  answerer  User  @relation(name: "answerer")
}
```
And then we have to readjust the column width for Question, which now doesn't match the column with of User and everything becomes harder to scan

```
model User {
  id        Int         @id
  asked     Question[]  @relation(name: "asker")
  answered  Question[]  @relation(name: "answerer")
}

model Question {
  id             Int   @id
  asker          User  @relation(name: "asker")
  answerer       User  @relation(name: "answerer")
  quiteLongName  Int
}
```

I understand the formatter should handle most of the writing problems for you but I know some part of me would still end up lining everything up myself rather than just letting the formatter handle it ;).

> The two constraints we're dealing with are that models don't look great this way...

I agree that model does not look good. The document gave me the impression that the configuration and the datamodel were separate creatures living in separate files but I understand the motivation more now so thank you for the explanation.

It's also looking better in general than when I last read it so good job there too ;)

**Comment by ejoebstl** [comment]
`@relation` attribute. I believe that the name should be the optionally named argument.

The name needs to be specified in a lot of cases, the related fields usually default to ID, except in rare corner cases.
